### PR TITLE
[SYCL][L0] Implementation of ext_oneapi_prod API

### PR DESCRIPTION
**Pull request by lbushi25**
This PR includes a preliminary implementation of the ext_oneapi_prod extension defined here:
https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_prod.asciidoc
It is preliminary because the UR-Level Zero queue flushing method which I've used to implement this extension is a no-op at the moment but I have a PR in review to change that: https://github.com/oneapi-src/unified-runtime/pull/1545.
Afterwards, my guess is that I will have to update the UR version we use in this repo to include that new commit once the Unified Runtime PR is merged. 
Nevertheless, I think I can continue with the review and merge of this PR independently of that since compilation will succeed because the UR flush method is defined, it just does nothing at the moment.

### CODE REVIEW

**Review by againull — COMMENTED** [review]
PR mentions only L0. Is this going to be supported for other backends?

File: sycl/source/detail/queue_impl.hpp
Comment (againull) [inline sycl/source/detail/queue_impl.hpp:393]: I think it is possible for MQueues to be empty. At least internally we most likely use  host queues which don't have any PI handles.

So, it will be nice to have some tests for this feature.

**Comment by lbushi25** [comment]
> PR mentions only L0. Is this going to be supported for other backends?

I dont know whether it will be supported for other backends. From what I can see it is already implemented for OpenCL but not HIP and CUDA.

File: sycl/source/detail/queue_impl.hpp
Comment (againull) [inline sycl/source/detail/queue_impl.hpp:393]: Also, I think special handling is needed for the case when graph recording is in progress. (for example, see queue_impl::wait).  It will be nice to have test for this case as well.

Comment (lbushi25) [reply sycl/source/detail/queue_impl.hpp:393]: Ok, I'll add a check to verify that MQueues is not empty. I'm also working on adding some tests on this PR.

File: sycl/source/feature_test.hpp.in
Comment (lbushi25) [reply sycl/source/feature_test.hpp.in:107]: I'll wait until the UR changes are merged and then merge this one.

**Comment by lbushi25** [comment]
@againull I added a check for when the MQueues member is empty and added some code to throw an exception when the queue is in graph recording mode since I guess it does not make sense to call flush while commands are being recorded.

File: sycl/source/detail/queue_impl.hpp
Comment (againull) [inline sycl/source/detail/queue_impl.hpp:392]: We should explicitly indicate here that this is hint only, because this gives a wrong impression that it is guaranteed that all previous commands will start executing.

I believe that there can be situations in SYCL scheduler when command which is submitted to the scheduler is not enqueued to the PI queue right away (if this enqueue is blocked by something, most like this can happen when host tasks are involved), so piQueueFlush will not flush such command.

Since this is a hint only per documentation and this only preliminary implementation, I believe just calling piQueueFlush is fine.

**Review by againull — COMMENTED** [review]
As I have mentioned earlier, If you consider this feature as supported, then you should move  documentation to supported directory.

File: sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc:1]: It looks like you copied the specification file to "doc/extensions/supported" but did not remove the old copy in "doc/extensions/proposed".  Can you remove the old copy?  Git should then show the file as "moved", and it will show any changes as diffs.

File: sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc:23]: Update the date.

File: sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc:49]: This status section should be updated as described in [README-process](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/README-process.md#implementing-an-extension).

File: sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc:39]:
```suggestion
This extension is written against the SYCL 2020 revision 8 specification.  All
```

Can you update this to the latest revision?

**Review by againull — APPROVED** [review]

**Review by gmlueck — APPROVED** [review]
Spec changes LGTM.

File: sycl/source/detail/queue_impl.hpp
Comment (RaviNarayanaswamy) [inline sycl/source/detail/queue_impl.hpp:391]: If this is not a guarantee then it is useless users like OpenMP will be assuming the job is submitted and inserting barriers to add more dependent tasks.

Comment (lbushi25) [reply sycl/source/detail/queue_impl.hpp:391]: I'm not sure I follow. The spec of the extension simply says that eventual execution is not guaranteed by calling ext_oneapi_prod(). Whether an app uses or not this API should not affect the correctness of the program. It is merely an encouragement to increase performance. Perhaps @gmlueck can help interpret the spec and its intended usage if I'm misunderstanding!

Comment (gmlueck) [reply sycl/source/detail/queue_impl.hpp:391]: @lbushi25's interpretation of the spec is correct.

@RaviNarayanaswamy, I don't understand your concern with barriers and dependent tasks.  How does that relate to whether the commands submitted prior to the "prod" call are guaranteed to start executing?

The thing that "prod" does *not* guarantee is that the commands are necessarily running at the same time as the host thread.  For example, it would be invalid for the host thread to spin reading host memory waiting until the kernel wrote some value there.

Comment (RaviNarayanaswamy) [reply sycl/source/detail/queue_impl.hpp:391]: @gmlueck  when we call fflush  we want it to submit the previous commands before returning.

Comment (lbushi25) [reply sycl/source/detail/queue_impl.hpp:391]: @RaviNarayanaswamy But I don't think there is a need for that. The queue class itself is an abstraction that allows you to submit commands and get an event back for synchronization. From your end you can think of the kernel as submitted. The flush functionality in this PR is concerned with encouraging execution of the kernel on the device itself, a lower level than what SYCL exposes..

Comment (RaviNarayanaswamy) [reply sycl/source/detail/queue_impl.hpp:391]: @lbushi25  is this PR implementing the feature requested in [CMPLRLLVM-37616].  If so OpenMP which calls MKL which submits the kernel and returns to OpenMP.  Now OpenMP needs to make sure the command is submitted,  we cannot use SYCL events to order the scheduling in OpenMP.  
If this is not for supporting the above JIRA and just ignore my comments.

Comment (lbushi25) [reply sycl/source/detail/queue_impl.hpp:391]: @RaviNarayanaswamy Based on internal discussion, it seems like the prod API is indeed sufficient for your use-case. Can you confirm? 
If so, please leave a review/approval.

**Review by RaviNarayanaswamy — APPROVED** [review]

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (AlexeySachkov) [inline sycl/plugins/unified_runtime/CMakeLists.txt:97]: These changes should be reverted before we merge the PR, or this URL & hash should be updated to point to the main UR branch in the main repo.

I don't think that this PR has a strict dependency on the UR. Since the API you are implementing is a hint, completely ignoring it technically a valid implementation. The fact that `urQueueFlush` does nothing for L0 backend is a UR bug, not SYCL RT bug

File: sycl/source/feature_test.hpp.in
Comment (AlexeySachkov) [reply sycl/source/feature_test.hpp.in:107]: This feature is a hint and implementation may decide to ignore it completely. Therefore, I think that it is fine to say that this is supported, even if we have some bugs on UR side

File: sycl/source/detail/queue_impl.hpp
Comment (AlexeySachkov) [inline sycl/source/detail/queue_impl.hpp:401]: Why do we only flush the first queue? What about other queues? Do we even have a practical case when SYCL queue will encapsulate several underlying queues?

File: sycl/unittests/Extensions/OneAPIProd.cpp
Comment (AlexeySachkov) [inline sycl/unittests/Extensions/OneAPIProd.cpp:2]:
```suggestion
//==-------- OneAPIProd.cpp --- sycl_ext_oneapi_prod unit tests ------------==//
```

File: sycl/unittests/Extensions/OneAPIProd.cpp
Comment (AlexeySachkov) [inline sycl/unittests/Extensions/OneAPIProd.cpp:37]: I think that it is better to use [`FAIL`](https://google.github.io/googletest/reference/assertions.html#FAIL) directly. You should also be able to do something like:
```c++
FAIL() << "exception is expected when calling `ext_oneapi_prod` while graph is being recorded";
```

You can also use [`EXPECT_THROW`](https://google.github.io/googletest/reference/assertions.html#EXPECT_THROW), but it won't allow you to validate error message or error code associated with the exception

File: sycl/unittests/Extensions/OneAPIProd.cpp
Comment (AlexeySachkov) [inline sycl/unittests/Extensions/OneAPIProd.cpp:39]: I would suggest to validate error code and maybe error message to ensure that you have caught the expected exception

**Review by AlexeySachkov — DISMISSED** [review]
I'm only requesting changes to prevent accidental merges because of custom URL used to UR tag.

File: sycl/source/detail/queue_impl.hpp
Comment (lbushi25) [reply sycl/source/detail/queue_impl.hpp:401]: Thanks, missed this, the comment next to the declaration of MQueues seems to suggest that multiple queues are neede for FPGA devices.

File: sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_oneapi_prod.asciidoc:23]: This looks like a merge mistake.  PR #13623 changed the copyright date to a single year.

**Review by kbenzie — APPROVED** [review]
UR LGTM

**Pull request merged**
